### ISSUE TITLE
Proper handling of colors in Python

### ISSUE THREAD

**Issue by Ecco** [numworks/epsilon#1533]
#### Problem you'd like to fix

Specifying a color in Python is lacking:
* We don't support named colors
* We don't support float-based ranges
* It's been re-implemented differently in `mod_kandinsky`, `mod_turtle`, and `mod_matplotlib`.

#### Describe the solution you'd like

Implement a unified `parseColor` function that lives in Python only (we specifically don't want to use those features outside of Python), and then use that function from the 3 aforementioned modules. The function should be defined in `python/port/port.(h/cpp)` with the following signature:

```c++
namespace MicroPython{

enum class ColorMode {
  MaxIntensity1,
  MaxIntensity255
};

KDColor MicroPython::ParseColor(mp_obj_t input, ColorMode mode);
}
```

It should be able to parse the following parameters:
|Type|Value|ColorMode|
|-|-|-|
|String| `"#FF0000"`|Any|
|String|`"#ff0000"`|Any|
|String|`0.5`|1|
|String| `"red"`|Any|
|String| `"r"` | Any |
|Tuple of int/float | `(1.0, 0.0, 0.0)` | 1 |
|Tuple of int/float| `(255, 0, 0)` | 255 |

#### Notes

* As far as we know, neither Turtle nor Matplotlib support passing a color from a single int (e.g. `0xFF0000`) or float (e.g. `0.5`) value, so we're better off not do this for now.
* The ColorMode parameter should always be 1 in `modmatplotlib`, always be 255 in `modkandinsky`, and should be configurable in `modturtle` (through a `colormode` function) with a default value of 1.

**Comment by RedGl0w** [comment]
I don't exactly understand what's the `0.5` string is
By the way, you could also support int as colors, because if you add `"#ffffff` why not adding `0xffffff`?

**Comment by Ecco** [comment]
> I don't exactly understand what's the `0.5` string is

It's a mid-grey. See [matplotlib's spec](https://matplotlib.org/tutorials/colors/colors.html).

> By the way, you could also support int as colors, because if you add `"#ffffff` why not adding `0xffffff`?

See the note above 😄 It's not supported elsewhere so let's not bother for now 😄

**Comment by RedGl0w** [comment]
> ```c++
> namespace MicroPython{
> 
> enum class ColorMode {
>   MaxIntensity1,
>   MaxIntensity255
> };
> ```

A `ColorMode::MaxIntensityAuto` could be very usefull for finding automaticaly the good ColorMode, if it wasn't set

**Comment by Ecco** [comment]
> A `ColorMode::MaxIntensityAuto` could be very usefull for finding automaticaly the good ColorMode, if it wasn't set

I think that's not possible. And that's actually the whole point of this enum: if an accurate auto-detection was possible, we might as well just drop the enum.

For instance, what color would `(1.0, 0, 0)` represent? Then how about `(123.0, 0, 0)`? I think it's impossible to make an automatic choice that's also consistent. Hence the enum 😄

**Comment by RedGl0w** [comment]
> For instance, what color would `(1.0, 0, 0)` represent? Then how about `(123.0, 0, 0)`? I think it's impossible to make an automatic choice that's also consistent. Hence the enum smile

If it is a float tuple, and if the values are < 1.0, it is a color mode 1, else it is a color mode 255

**Comment by Ecco** [comment]
But then drawing a gradient using this syntax will be completely f*cked up when you get close to 1 😄

**Comment by RedGl0w** [comment]
> But then drawing a gradient using this algorithm will be completely f*cked up when you get close to 1 smile

Euh... No ? Why should it be messed up ? If the 1.0 value is the issue, why not just not supporting float > 1.0 values ?

**Comment by Ecco** [comment]
> Why should it be messed up ?

Let's say a user writes a script to draw a 200-steps gradient using this algorithm, from `(100.0,100.0,100.0)` to `(0.0,0.0,0.0)`. Using an auto-detection algorithm like the one you describe would indeed draw a gradient from mid-gray to black, but the last 3 pixels would be full white, mid gray, black, instead of mostly black, mostly black, black.

> If the 1.0 value is the issue, why not just not supporting float > 1.0 values ?

Because again it wouldn't be consistent. In the example above then what would happen for the last 3 pixels?

No other library does it anyway, so I think it's better to just either pick one convention (255 for Kandinsky, 1 for Matplotlib, or let the user decide (turtle)).

**Comment by RedGl0w** [comment]
Ok I just realized that my idea was breaking maybe some scripts, so never mind :|

**Comment by RedGl0w** [comment]
> | Type   | Value | ColorMode |
> |--------|-------|-----------|
> | String | "r"   | any       |

How will it be interpreted with `b` ? Or will it be interpreted only with colors which doesn't start with the same letter ?

**Comment by Ecco** [comment]
> How will it be interpreted with b ?

Matplotlib's [spec](https://matplotlib.org/tutorials/colors/colors.html) is a good start.

In practice, it's just another set of named colors. So we won't need any special algorithm for those, which is the reason we added them to the list.

**Comment by EmilieNumworks** [comment]
Matplotlib is yet to be implemented!

**Comment by artaxxx** [comment]
Fixed in 14.4.0
